File: samples/java-mule/src/main/scala/com/wordnik/swagger/mule/ApiListingJSON.scala
package com.wordnik.swagger.mule

import com.wordnik.swagger.core.util.ReaderUtil
import com.wordnik.swagger.annotations._
import com.wordnik.swagger.jaxrs.config.BeanConfig
import com.wordnik.swagger.core.filter._
import com.wordnik.swagger.config._
import com.wordnik.swagger.model._
import com.wordnik.swagger.reader._
import com.wordnik.swagger.core.util.JsonSerializer

import org.slf4j.LoggerFactory

import javax.ws.rs._
import javax.ws.rs.core.{ Response, MediaType }

@Path("/api-docs")
class ApiListingJSON extends ApiListing {
  @GET
  @Produces(Array(MediaType.APPLICATION_JSON))
  def getResourceJson = {
    val o = getResource()
    println(JsonSerializer.asJson(o))
    Response.ok(JsonSerializer.asJson(o)).build()
  }

  @GET
  @Produces(Array(MediaType.APPLICATION_XML))
  def getResourceXml = Response.ok(JsonSerializer.asXml(A(getResource()))).build()

  @GET
  @Produces(Array(MediaType.APPLICATION_JSON))
  @Path("/{route: .+}")
  def apiDeclarationJson (
    @PathParam("route") route: String
  ): Response = {
    val listings = getApiDeclaration(route)
    listings.size match {
      case 1 => Response.ok( JsonSerializer.asJson( listings(0))).build
      case _ => Response.status(404).build
    }
  }

  @GET
  @Produces(Array(MediaType.APPLICATION_XML))
  @Path("/{route: .+}")
  def apiDeclarationXml (
    @PathParam("route") route: String
  ): Response = {
    val listings = getApiDeclaration(route)
    listings.size match {
      case 1 => Response.ok(JsonSerializer.asXml(B(listings(0)))).build
      case _ => Response.status(404).build
    }
  }
}

case class A(resourceListing: ResourceListing)
case class B(apiDeclaration: com.wordnik.swagger.model.ApiListing)

trait ApiListing {
  private val LOGGER = LoggerFactory.getLogger(classOf[ApiListing])

  def getResource(): ResourceListing = {
    val docRoot = ""
    val queryParams = Map[String, List[String]]()
    val cookies = Map[String, String]()
    val headers = Map[String, List[String]]()

    val f = new SpecFilter
    val listings = ApiListingCache.listing(docRoot).map(specs => {
      (for(spec <- specs.values) 
        yield f.filter(spec, FilterFactory.filter, queryParams, cookies, headers)
      ).filter(m => m.apis.size > 0)
    })
    val references = (for(listing <- listings.getOrElse(List())) yield {
      ApiListingReference(listing.resourcePath, listing.description)
    }).toList
    val config = ConfigFactory.config
    ResourceListing(config.apiVersion,
      config.swaggerVersion,
      references
    )
  }

  def getApiDeclaration(route: String) = {
    LOGGER.debug("requested apiDeclaration for " + route)
    val docRoot = this.getClass.getAnnotation(classOf[Path]).value
    val queryParams = Map[String, List[String]]()
    val cookies = Map[String, String]()
    val headers = Map[String, List[String]]()
    val f = new SpecFilter
    val pathPart = cleanRoute(route)
    LOGGER.debug("requested route " + pathPart)
    ApiListingCache.listing(docRoot).map(specs => {
      (for(spec <- specs.values) yield {
        LOGGER.debug("inspecting path " + spec.resourcePath)
        f.filter(spec, FilterFactory.filter, queryParams, cookies, headers)
      }).filter(m => {
        val resourcePath = m.resourcePath match {
          case e: String if(e.startsWith("/")) => e
          case e: String => "/" + e
        }
        resourcePath == pathPart
      })
    }).toList.flatten
  }

  // ensure leading slash, remove trailing
  def cleanRoute(route: String) = {
    val cleanStart = {
      if(route.startsWith("/")) route
      else "/" + route
    }
    if(cleanStart.endsWith("/")) cleanStart.substring(0, cleanStart.length - 1)
    else cleanStart
  }
}

object ApiListingCache extends ReaderUtil {
  var cache: Option[Map[String, com.wordnik.swagger.model.ApiListing]] = None

  def listing(docRoot: String): Option[Map[String, com.wordnik.swagger.model.ApiListing]] = {
    cache.orElse{
      ClassReaders.reader.map{reader => 
        ScannerFactory.scanner.map(scanner => {
          val classes = scanner match {
            case scanner: Scanner => {
              val beanConfig = scanner.asInstanceOf[BeanConfig]
              beanConfig.classesFromContext(null, null)
            }
            case _ =>List()
          }
          val listings = (for(cls <- classes) yield reader.read(docRoot, cls, ConfigFactory.config)).flatten
          val mergedListings = groupByResourcePath(listings)
          cache = Some(mergedListings.map(m => (m.resourcePath, m)).toMap)
        })
      }
      cache
    }
  }
}

File: modules/swagger-mule/src/main/java/com/wordnik/swagger/mule/ApiListingJSON.java
package com.wordnik.swagger.mule;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

import javax.servlet.FilterConfig;
import javax.servlet.ServletConfig;
import javax.servlet.ServletContext;
import javax.ws.rs.GET;
import javax.ws.rs.Path;
import javax.ws.rs.Produces;
import javax.ws.rs.core.Application;
import javax.ws.rs.core.Context;
import javax.ws.rs.core.Cookie;
import javax.ws.rs.core.HttpHeaders;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.MultivaluedMap;
import javax.ws.rs.core.Response;
import javax.ws.rs.core.UriInfo;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.sun.jersey.spi.container.servlet.WebConfig;
import com.wordnik.swagger.config.FilterFactory;
import com.wordnik.swagger.config.Scanner;
import com.wordnik.swagger.config.ScannerFactory;
import com.wordnik.swagger.config.SwaggerConfig;
import com.wordnik.swagger.core.filter.SpecFilter;
import com.wordnik.swagger.core.filter.SwaggerSpecFilter;
import com.wordnik.swagger.jaxrs.Reader;
import com.wordnik.swagger.jaxrs.config.JaxrsScanner;
import com.wordnik.swagger.jaxrs.listing.SwaggerSerializers;
import com.wordnik.swagger.models.Swagger;

@Path ("/api-docs")
@Produces (MediaType.APPLICATION_JSON)
public class ApiListingJSON {
	private static final Logger	LOGGER		= LoggerFactory.getLogger(ApiListingJSON.class);
	static boolean				initialized	= false;

	protected synchronized Swagger scan(Application app, ServletConfig sc) {
		Swagger swagger = null;
		Scanner scanner = ScannerFactory.getScanner();
		LOGGER.debug("using scanner " + scanner);
		if (scanner != null) {
			SwaggerSerializers.setPrettyPrint(scanner.getPrettyPrint());
			swagger = (Swagger) context.getAttribute("swagger");
			Set<Class<?>> classes = null;
			if (scanner instanceof JaxrsScanner) {
				JaxrsScanner jaxrsScanner = (JaxrsScanner) scanner;
				classes = jaxrsScanner.classesFromContext(app, sc);
			} else {
				classes = scanner.classes();
			}
			if (classes != null) {
				Reader reader = new Reader(swagger);
				swagger = reader.read(classes);
				if (scanner instanceof SwaggerConfig)
					swagger = ((SwaggerConfig) scanner).configure(swagger);
				else {
					SwaggerConfig configurator = (SwaggerConfig) context.getAttribute("reader");
					if (configurator != null) {
						LOGGER.debug("configuring swagger with " + configurator);
						configurator.configure(swagger);
					} else LOGGER.debug("no configurator");
				}
				context.setAttribute("swagger", swagger);
			}
		}
		initialized = true;
		return swagger;
	}

	// Doesn't work
	@Context
	ServletContext	context;

	@GET
	public Response getListingJson(@Context Application app, /* Also doesn't work: @Context ServletConfig sc, */
			@Context HttpHeaders headers, @Context UriInfo uriInfo) {
		//		Swagger swagger = (Swagger) context.getAttribute("swagger");
		//		if (!initialized)
		//		Swagger swagger = scan(app, sc);
		//		if (swagger != null) {
		//			SwaggerSpecFilter filterImpl = FilterFactory.getFilter();
		//			if (filterImpl != null) {
		//				SpecFilter f = new SpecFilter();
		//				swagger = f.filter(swagger, filterImpl, getQueryParams(uriInfo.getQueryParameters()),
		//						getCookies(headers), getHeaders(headers));
		//			}
		//			return Response.ok().entity(swagger).build();
		//		} else return Response.status(404).build();
		return Response.ok("Hi").build();
	}

	protected Map<String, List<String>> getQueryParams(MultivaluedMap<String, String> params) {
		Map<String, List<String>> output = new HashMap<String, List<String>>();
		if (params != null) {
			for (String key : params.keySet()) {
				List<String> values = params.get(key);
				output.put(key, values);
			}
		}
		return output;
	}

	protected Map<String, String> getCookies(HttpHeaders headers) {
		Map<String, String> output = new HashMap<String, String>();
		if (headers != null) {
			for (String key : headers.getCookies().keySet()) {
				Cookie cookie = headers.getCookies().get(key);
				output.put(key, cookie.getValue());
			}
		}
		return output;
	}

	protected Map<String, List<String>> getHeaders(HttpHeaders headers) {
		Map<String, List<String>> output = new HashMap<String, List<String>>();
		if (headers != null) {
			for (String key : headers.getRequestHeaders().keySet()) {
				List<String> values = headers.getRequestHeaders().get(key);
				output.put(key, values);
			}
		}
		return output;
	}
}
